compiler/src/dotty/tools/dotc/transform/LazyVals.scala

```diff
@@ -274,6 +284,274 @@ class LazyVals extends MiniPhase with IdentityDenotTransformer {
+   * ```
+   * private @volatile var _x: AnyRef = null
+   * @tailrec def x: A =
+   *   _x match
+   *   case current: A => 
+   *     current
+   *   case NullValue => null
+   *   case null =>
+   *     if CAS(_x, null, Evaluating) then
+   *       var result: AnyRef = null // here, we need `AnyRef` to possibly assign `NullValue`
+   *       try
+   *         result = rhs
+   *         nullable = null // if the field is nullable; see `CollectNullableFields`
+   *       finally
+   *         if result == null then result = NullValue // drop if A is non-nullable
+   *         if !CAS(_x, Evaluating, result) then
+   *           val lock = _x.asInstanceOf[Waiting]
+   *           CAS(_x, lock, result)
+   *           lock.release()
+   *     x
+   *   case Evaluating =>
+   *     CAS(_x, Evaluating, new Waiting)
+   *     x
+   *   case current: Waiting =>
+   *     current.awaitRelease()
+   *     x
+   * ```
+   * Where `Evaluating` and `NullValue` are represented by `object`s and `Waiting` by a class that
+   * allows awaiting the completion of the evaluation. Note that since tail-recursive
+   * functions are transformed *before* lazy-vals, this implementation directly implements
+   * the resulting loop. `PatternMatcher` coming before `LazyVals`, the pattern matching block
+   * is implemented using if-s. Additionally, the code can be optimized to be better inlined by 
+   * introducing separate function for computing the lazy value and awaiting it. That is:
+   * 
+   * ```
+   * private var _x: AnyRef = null
+   * 
+   * def x: A =
+   *   if !(_x == null || _x.isInstanceOf[LazyValControlState]) then
+   *     return _x.asInstanceOf[A]
+   *   else if _x == NullValue then
+   *     return null
+   *   else 
+   *     return x_compute()
+   * 
+   * private def x_compute() = 
+   *   while <EmptyTree> do
+   *     val current: AnyRef = _x
+   *     if current == null then
+   *       if CAS(_x, null, Evaluating) then
+   *         var resultNullable: AnyRef = null
+   *         var result: AnyRef = null
+   *         try
+   *           resultNullable = rhs
+   *           nullable = null
+   *           if resultNullable == null then
+   *             result = NullValue
+   *           else
+   *             result = resultNullable
+   *           return result
+   *         finally
+   *            if !CAS(_x, Evaluating, result) then
+   *              val lock = _x.asInstanceOf[Waiting]
+   *              CAS(_x, lock, result)
+   *              lock.release()
+   *     else
+   *       if current.isInstanceOf[LazyValControlState] then
+   *         if current.isInstanceOf[Evaluating] then // To avoid creating Waiting instance
+   *           CAS(current, Evaluating, new Waiting)
+   *         else if current.isInstanceOf[NullValue] then
+   *           return null
+   *         else if current.isInstanceOf[Waiting] then
+   *           current.asInstanceOf[Waiting].await()
```

Comment:
This does not correspond to what's actually emitted anymore. (`isInstanceOf` vs `==`, or `eq`, and the order).